Create a generic number ↔ name and let EVP use it

levitte opened the pull request:
This creates a generic, per library context number ↔ name map, and has the EVP fetcher use it.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated

levitte committed:
Create internal number<->name mapping API

This can be used as a general name to identity map.

levitte committed:
Make the generic EVP fetching mechanism use the namenum map

This avoids using the ASN1_OBJECT database, which is bloated for the
purpose of a simple number <-> name database.

levitte commented:
Arguably, this namemap could be used by the property code...  I don't quite understand why `crypto/property/property_string.c` currently holds two databases, when they are really just string ↔ identity maps and those strings don't have any other inherent per se.

paulidale commented:
@levitte the reason that `crypto/property/property_string.c` holds two (name, value) databases is to distinguish between the _names_ and the _values_.   There are strict rules about what forms _names_ can take and these are enforced.  There are far laxer restrictions on _values_.  Keeping them separate seemed expedient from the point of view of validity checking.

I'm not going to object to a merging of the two, so long as the semantics don't also change.

mattcaswell commented on crypto/core_namemap.c:
```diff
@@ -0,0 +1,156 @@
+
+    strcpy(entry->body, name);
+    entry->name = entry->body;
+    entry->number = sk_NAMEMAP_push(store->numname, entry);
```
Doesn't this need to be locked?

levitte replied:
Yes, and I will add that. I was very short on time when I put together this PR...

paulidale commented on crypto/core_namemap.c:
```diff
@@ -0,0 +1,156 @@
+} NAMEMAP;
+
+DEFINE_LHASH_OF(NAMEMAP);
+DEFINE_STACK_OF(NAMEMAP)
```
Sparse array instead here?
This kind of lookup is why it was added.

levitte replied:
Sure...

levitte replied:
Your discussion comments make me wonder, though, if you have the right idea about the code implemented here...

paulidale replied:
I'm wrong :(

paulidale commented:
There is a slight difference between the properties name ↔ value maps and this.  The property maps are known to be indexed sequentially (from zero) which makes a stack of names efficient, here the index doesn't have that property which makes a sparse array the better choice.

By switching the property implementation over, there will be a (small?) space cost but I don't think there will be a noticeable time cost.  I believe the simplification to a single name ↔ value map across the library is a beneficial one that outweighs associated the costs.

paulidale commented:
Merging the two name ↔ value maps in the property code into a single map will allow properties like: "fips=foo, foo=bar" (where _foo_ is **not** a property name) to be specified and parsed without error.  I.e. the reservation of names without a period can be circumvented.

levitte commented:
> There is a slight difference between the properties name ↔ value maps and this. The property maps are known to be indexed sequentially (from zero) which makes a stack of names efficient, here the index doesn't have that property which makes a sparse array the better choice.

Uhmmmm, this namemap is indexed monotonically as well, but from one (zero indicates an "undefined" string).  A string gets added by pushing it in the stack...

levitte commented:
There's one bit that I find highly disturbing, and it's that this namemap is per library context...  That would mean duplicating it unnecessarily if we're handling several library contexts within a unit.  It's just a string to number cache, essentially, so there's no reason to make the numbers different per context.

I *could* force it to end up in the default library context, but that would mean that there must be one, and the FIPS module currently forbids that...  I might have an idea or two on that, btw, but that's the matter of another PR.

paulidale commented:
I saw the use of `sk_NAMEMAP_value` and thought that it searched for some unknown reason.

Best to stick with the stack and it can directly replace the two property mappings.

paulidale commented:
Could we have this per context and default to the "default" context if none is directly set?

paulidale commented:
Properties index from 1 as well.

levitte commented:
>  Could we have this per context and default to the "default" context if none is directly set?

We can.  Also, this implementation only allows one namemap (per library context), and what you said about keeping property names and values separate makes me think that we need a mechanism to have more than one.  This is perfectly viable, but needs an extension of the API.

levitte committed:
fixup! Create internal number<->name mapping API

levitte committed:
fixup! Create internal number<->name mapping API

levitte committed:
fixup! Create internal number<->name mapping API

levitte committed:
fixup! Make the generic EVP fetching mechanism use the namenum map

levitte commented:
I re-implemented it to make it possible to create free floating name maps.

levitte commented:
Ooooh, this found me an interesting bug...  we weren't doing things quite right when there is no pre-existing nid for a name...

levitte committed:
EVP_FETCH: deal with names without pre-defined NIDs

We didn't deal very well with names that didn't have pre-defined NIDs,
as the NID zero travelled through the full process and resulted in an
inaccessible method.  By consequence, we need to refactor the method
construction callbacks to rely more on algorithm names.

We must, however, still store the legacy NID with the method, for the
sake of other code that depend on it (for example, CMS).

levitte commented:
This isn't in draft any more

levitte committed:
fixup! Create internal number<->name mapping API

levitte committed:
util/find-doc-nits: Fine tune detection of POD markup in NAME section

POD markup is only forbidden in the actual names, while permitted in
the description.

slontis commented on crypto/evp/evp_fetch.c:
```diff
@@ -112,30 +113,13 @@ static void *construct_method(const char *algorithm_name,
+    int nid;
 
-    if (nid == NID_undef)
+    if ((nid = ossl_namemap_number(methdata->libctx, algorithm_name)) == 0
```
The namemap_new seems to call ossl_namemap_number() and early out if its already mapped - so you may not need the first line.

levitte replied:
Things have changed since this comment...  care to take a refreshed look?

slontis commented on crypto/core_namemap.c:
```diff
@@ -0,0 +1,217 @@
+
+struct ossl_namemap_st {
+    /* Flags */
+    unsigned int floating:1; /* If 0, it's stored in a library context */
```
Not sure about the name - maybe something alloc related would be better.

levitte replied:
Well, it *is* allocated either way.  "stored"?

slontis replied:
yep sounds better

slontis commented on crypto/core_namemap.c:
```diff
@@ -0,0 +1,217 @@
+{
+    NAMEMAP_ENTRY *entry;
+
+#if 0                            /* TODO(3.0) */
```
Is this a #ifndef FIPS_MODE

levitte replied:
No.  But FIPS module related changes I have in my mind will hopefully make this possible, and enabled at all times, unconditionally...  if it works out like I imagine (there will be another PR on this)

mattcaswell replied:
I'd prefer not to see this "if 0" code here. I think its better to add it in whatever future PR enables this functionality. It doesn't really add anything here since I can't see the big picture of how you intend to use this capability.

slontis commented on crypto/evp/evp_fetch.c:
```diff
@@ -146,21 +135,22 @@ static void destruct_method(void *method, void *data)
 }
 
 void *evp_generic_fetch(OPENSSL_CTX *libctx, int operation_id,
-                        const char *algorithm, const char *properties,
```
why was this changed to name?

levitte replied:
For consistency.  We have "name", "algorithm_name" and "algorithm".  Since this is EVP, we *know* they are names of algorithms, and this parameter *is* the algorithm name, so...

slontis commented on doc/internal/man3/ossl_namemap_new.pod:
```diff
@@ -0,0 +1,81 @@
+ void ossl_namemap_free(OSSL_NAMEMAP *namemap);
+
+ int ossl_namemap_add(OSSL_NAMEMAP *namemap, const char *name);
+ const char *ossl_namemap_name(OSSL_NAMEMAP *namemap, int number);
```
const OSSL_NAMEMAP ?

levitte replied:
Good point

slontis replied:
and then next one too

slontis commented on doc/internal/man3/ossl_namemap_new.pod:
```diff
@@ -0,0 +1,81 @@
+=head1 DESCRIPTION
+
+A B<OSSL_NAMEMAP> is a simple number E<lt>-E<gt> name map, which can
+be used to give arbitrary any name (any string) a unique dynamic
```
any arbitrary

levitte replied:
Okie

slontis commented on doc/internal/man3/ossl_namemap_new.pod:
```diff
@@ -0,0 +1,81 @@
+This is suitable to use when the B<OSSL_NAMEMAP> is embedded in other
+structures, or should be independent for any reason.
+
+ossl_namemap_stored() finds or a auto-creates the default namemap in
```
remove 'a'

levitte replied:
Sure

slontis commented on include/internal/namemap.h:
```diff
@@ -0,0 +1,23 @@
+const char *ossl_namemap_name(OSSL_NAMEMAP *namemap, int number);
+int ossl_namemap_number(OSSL_NAMEMAP *namemap, const char *name);
+
+
```
nit space

levitte replied:
Good point

levitte committed:
fixup! Create internal number<->name mapping API

paulidale commented on crypto/core_namemap.c:
```diff
@@ -0,0 +1,217 @@
+
+    if (namemap != NULL) {
+        namemap->stored = 1;
+    }
```
Lose the {}s ?

levitte replied:
Done

paulidale commented on crypto/core_namemap.c:
```diff
@@ -0,0 +1,217 @@
+{
+    OSSL_NAMEMAP *namemap = vnamemap;
+
+    /* Pretend it's stored, or ossl_namemap_free() will do nothing */
```
Is this meant to say _not_ stored?

levitte replied:
Right.  Fixed

paulidale commented on crypto/core_namemap.c:
```diff
@@ -0,0 +1,217 @@
+}
+
+/*
+ * TODO(3.0) Make NULL to signify the default namemap, found in the default
```
I wonder if this is the best place to convert NULL to default?

- If the property string storage is modified to use this data structure, it would be bad.
- It would be okay using the property string data structure here, I think.
- The third obvious choice is to abstract the data structure and share it between both places.

levitte replied:
Since the property strings need to be distinguished from all other strings, for reasons you've already explained, I would say that you won't use NULL, but rather allocate your own namemaps as part of this structure:

https://github.com/openssl/openssl/blob/9acbe07d2300d34a7ea846d9756f33b4595e32fb/crypto/property/property_string.c#L37-L42

In other words, the structure could be remade like this:

``` C
typedef struct {
    OSSL_NAMEMAP *prop_names;
    OSSL_NAMEMAP *prop_values;
} PROPERTY_STRING_DATA;
```

... and the property code would only keep the higher end strings functions.

mattcaswell commented on crypto/core_namemap.c:
```diff
@@ -0,0 +1,217 @@
+        && (namemap->numname = sk_NAMEMAP_ENTRY_new_null()) != NULL
+        && (namemap->namenum =
+            lh_NAMEMAP_ENTRY_new(namemap_hash, namemap_cmp)) != NULL) {
+        namemap->stored = 0;
```
Not strictly necessary due to the zalloc

levitte replied:
Removed

mattcaswell commented on crypto/evp/evp_fetch.c:
```diff
@@ -68,16 +69,23 @@ static OSSL_METHOD_STORE *get_default_method_store(OPENSSL_CTX *libctx)
 
-    (void)ossl_method_store_fetch(store, methdata->nid, propquery, &method);
+    if ((namemap = ossl_namemap_stored(libctx)) == NULL
+        || (nid = ossl_namemap_add(namemap, name)) == 0)
```
While this is a nid in the sense that nid == numeric identifier, it is a different type of nid to where that term is used everywhere else in the code. I think that's confusing so a different name would be a good idea.

levitte replied:
I can certainly go with `id` if that's preferable.

mattcaswell replied:
Yes, I think that would be less confusing.

levitte replied:
Okie

mattcaswell commented on crypto/evp/evp_fetch.c:
```diff
@@ -107,35 +117,14 @@ static int put_method_in_store(OPENSSL_CTX *libctx, void *store,
-        tmpobj.length = 0;
-        tmpobj.data = NULL;
+    /* TODO(3.0) get rid of the need for legacy NIDs */
+    int legacy_nid = OBJ_sn2nid(name);
```
Why is this needed? OBJ_sn2nid uses ASN1_OBJECT and I thought removing that dependency that was part of the point of this PR?

levitte replied:
The legacy NID needs to be preserved because of things like this:

https://github.com/openssl/openssl/blob/9acbe07d2300d34a7ea846d9756f33b4595e32fb/crypto/evp/evp_lib.c#L25-L37

levitte replied:
Of course, I answered a bit too quickly, and only your first question.  I'll take a closer look at what I just quoted.

mattcaswell replied:
The above code only needs to work for implicit lookup ciphers so shouldn't impact the evp_fetch code.

levitte committed:
fixup! Create internal number<->name mapping API

levitte commented on crypto/evp/evp_fetch.c:
```diff
@@ -107,35 +117,23 @@ static int put_method_in_store(OPENSSL_CTX *libctx, void *store,
+     * functionality that use them.
+     */
+    legacy_nid = OBJ_sn2nid(name);
+#endif
```
So here's a point that may make this PR controversial.  I would like to propose that *within the FIPS module*, the "legacy" NIDs are (or should be) irrelevant.

In the particular case I identified where "legacy" NIDs are needed is all about ASN.1 parameters, something we won't have to deal with within the FIPS module as far as I understand.  I'll spend some time identifying others...

levitte replied:
... or perhaps not so controversial after all...  I just looked in #8728 and noticed that the "offending" code is excluded from the FIPS module, so...

mattcaswell commented on doc/internal/man3/ossl_namemap_new.pod:
```diff
@@ -0,0 +1,81 @@
+=head1 DESCRIPTION
+
+A B<OSSL_NAMEMAP> is a simple number E<lt>-E<gt> name map, which can
+be used to give arbitrary any arbitrary name (any string) a unique
```
arbitrary any arbitrary?

mattcaswell commented on doc/internal/man3/ossl_namemap_new.pod:
```diff
@@ -0,0 +1,81 @@
+It's possible that ossl_namemap_add(), ossl_namemap_name() and
+ossl_namemap_number() will change to accept NULL as a B<OSSL_NAMEMAP>
+to use one that is auto-created and associated with the default
+library context.
```
I'd delete the above. Either it does get added in the future - in which case we document it then. Or it doesn't, and this note becomes dead.

mattcaswell commented on util/find-doc-nits:
```diff
@@ -198,8 +199,6 @@ sub check()
     print "$id period in NAME section\n"
         if $contents =~ /=head1 NAME.*\.\n.*=head1 SYNOPSIS/ms;
-    print "$id POD markup in NAME section\n"
-        if $contents =~ /=head1 NAME.*[<>].*=head1 SYNOPSIS/ms;
```
Not sure I understand the purpose of this find-doc-nits change.

mattcaswell reviewed (changes requested):
Also the nid->id change is still outstanding

levitte committed:
EVP_FETCH: remove the need to transport the legacy NID through construction

Now that the legacy NID isn't used as a main index for fetched
algorithms, the legacy NID was just transported around unnecessarily.
This is removed, and the legacy NID is simply set by EVP_{API}_fetch()
after the construction process is done.

levitte committed:
fixup! EVP_FETCH: deal with names without pre-defined NIDs

levitte committed:
fixup! Create internal number<->name mapping API

levitte committed:
fixup! Create internal number<->name mapping API

levitte committed:
fixup! Create internal number<->name mapping API

levitte committed:
fixup! Create internal number<->name mapping API

mattcaswell commented on crypto/core_namemap.c:
```diff
@@ -0,0 +1,216 @@
+}
+
+/*
+ * TODO(3.0) Make NULL to signify the default namemap, found in the default
```
Right....that doesn't really answer the question. We're not using this capability as far as I can tell, so its not clear why its useful.

levitte replied:
Oh wait, yeahok, I see what you mean.  Fix coming up

levitte replied:
This is a generic namenum map.  With the current implementation, every library context will have its own default one, so for the string "foo", there will be as many entries, when the mapped numbers could just as well be exactly the same for all of them.

levitte committed:
fixup! Create internal number<->name mapping API

levitte commented:
Ping?

(btw, in a conversation yesterday, it was suggested that the strings could be bitstrings, i.e. binary data rather than just text...  I don't mind the idea, but I also think that could be future development of this map functionality, it doesn't have to happen in *this* PR)

levitte commented:
Doing the close+open thing to kick Appveyor

levitte commented:
Merged.

0211740fcc47a954be19cceb65fb57a6f7deb797 EVP_FETCH: remove the need to transport the legacy NID through construction
1f79ddf5049ff53ad8a7cbab76e62d02d9ac099f util/find-doc-nits: Fine tune detection of POD markup in NAME section
2e49c05472ab76cee4e30c2eaa4fa576b9ae92c6 EVP_FETCH: deal with names without pre-defined NIDs
baff732dc654f8aab37953f6036399989c765604 Make the generic EVP fetching mechanism use the namenum map
f2182a4e6fef098014117eb4a4f35ca84cd2cf8e Create internal number<->name mapping API
